Source/Dafny/Compilers/SinglePassCompiler.cs

```diff
@@ -4831,29 +4867,64 @@ private class GetterSetterLvalueImpl : ILvalue {
       }
     }
 
-    private ConcreteSyntaxTree MaybeInjectSubsetConstraint(ConcreteSyntaxTree wr, bool inLetExprBody, BoundVar bv,
-      Expression e, bool isReturning = false, bool elseReturnValue = false) {
-      if (bv.Type.NormalizeExpand(true) is UserDefinedType
-        {
-          TypeArgs: var typeArgs,
-          ResolvedClass:
-          SubsetTypeDecl
-          {
-            TypeArgs: var typeParametersArgs,
-            Var: var var,
-            ConstraintIsCompilable: true,
-            Constraint: var constraint
+    private ConcreteSyntaxTree MaybeInjectSubtypeConstraint(string tmpVarName,
+      Type collectionElementType, Type boundVarType, bool inLetExprBody,
+      Bpl.IToken tok, ConcreteSyntaxTree wr, bool isReturning = false, bool elseReturnValue = false
+      ) {
+      if (IsTargetSupertype(collectionElementType, boundVarType)) {
+        // That's an issue. We are iterating values that need to be checked.
```

Comment:
Instead of the comment, consider introducing a variable like 

`var iterationValuesNeedToBeChecked = IsTargetSupertype(collectionElementType, boundVarType)`